exercises/src/main/scala/fpinscala/monads/Monad.scala

```diff
@@ -34,25 +34,49 @@ trait Monad[M[_]] extends Functor[M] {
   def map2[A,B,C](ma: M[A], mb: M[B])(f: (A, B) => C): M[C] =
     flatMap(ma)(a => map(mb)(b => f(a, b)))
 
-  def sequence[A](lma: List[M[A]]): M[List[A]] = ???
-
-  def traverse[A,B](la: List[A])(f: A => M[B]): M[List[B]] = ???
-
-  def replicateM[A](n: Int, ma: M[A]): M[List[A]] = ???
-
-  def compose[A,B,C](f: A => M[B], g: B => M[C]): A => M[C] = ???
+  def sequence[A](lma: List[M[A]]): M[List[A]] =
+    lma.foldRight(unit(Nil:List[A]))((p, acc) => {
+      map2(p, acc)(_ :: _)
+    })
+
+  def traverse[A,B](la: List[A])(f: A => M[B]): M[List[B]] =
+    la.foldRight(unit(Nil:List[B]))((p, acc) => {
+      map2(f(p), acc)(_ :: _)
+    })
+
+  def replicateM[A](n: Int, ma: M[A]): M[List[A]] =
```

Comment:
This was my implementation:

```
  def replicateM[A](n: Int, ma: M[A]): M[List[A]] =
    sequence(List.fill(n)(ma))
```

It also has the advantage of being stack-safe as long as foldRight is.